In [1]:
import numpy as np


In [2]:
X = np.array([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
              [1, 1, 2, 1, 3, 0, 5, 10, 1, 2]])

In [3]:
y = [45, 55, 50, 59, 65, 35, 75, 80, 50, 60]


In [4]:
def calc_mse(y, y_pred):
    err = np.mean((y - y_pred)**2)
    return err

1. Подберите скорость обучения (alpha) и количество итераций:

In [5]:
n = X.shape[1]
alpha = 1e-4
W = np.array([1, 0.5])
W, alpha

(array([1. , 0.5]), 0.0001)

In [6]:
for i in range(200): 
    y_pred = np.dot(W, X)
    err = calc_mse(y, y_pred)    
    for ii in range(W.shape[0]):
        W[ii] -= alpha * (1/n * 2 * np.sum(X[ii] * (y_pred - y)))
    if i % 100 == 0:
        print(i, W, err)

0 [1.01102 0.534  ] 3173.15
100 [2.02168794 3.452835  ] 2207.6942101483005


In [7]:
n = X.shape[1]
alpha = 0.05
W = np.array([1, 0.5])
W, alpha

(array([1. , 0.5]), 0.05)

In [8]:
for i in range(200): 
    y_pred = np.dot(W, X)
    err = calc_mse(y, y_pred)    
    for ii in range(W.shape[0]):
        W[ii] -= alpha * (1/n * 2 * np.sum(X[ii] * (y_pred - y)))
    if i % 100 == 0:
        print(i, W, err)

0 [ 6.51 17.5 ] 3173.15
100 [47.03142979  3.94777749] 45.96159894380851


In [9]:
n = X.shape[1]
alpha = 0.06
W = np.array([1, 0.5])
W, alpha

(array([1. , 0.5]), 0.06)

In [10]:
for i in range(200): 
    y_pred = np.dot(W, X)
    err = calc_mse(y, y_pred)    
    for ii in range(W.shape[0]):
        W[ii] -= alpha * (1/n * 2 * np.sum(X[ii] * (y_pred - y)))
    if i % 100 == 0:
        print(i, W, err)

0 [ 7.612 20.9  ] 3173.15
100 [47.16625286  3.92288138] 45.94015500316733


In [11]:
n = X.shape[1]
alpha = 0.06
W = np.array([1, 0.5])
W, alpha

(array([1. , 0.5]), 0.06)

In [12]:
for i in range(200): 
    y_pred = np.dot(W, X)
    err = calc_mse(y, y_pred)    
    for ii in range(W.shape[0]):
        W[ii] -= alpha * (1/n * 2 * np.sum(X[ii] * (y_pred - y)))
    if i % 20 == 0:
        print(i, W, err)

0 [ 7.612 20.9  ] 3173.15
20 [35.85695572  6.15284752] 125.86627725452965
40 [44.0891315   4.49316871] 51.980145468359524
60 [46.3654434   4.07078732] 46.39687623882493
80 [46.99317184  3.95484256] 45.972423420722386
100 [47.16625286  3.92288138] 45.94015500316733
120 [47.21397542  3.91406904] 45.93770184282277
140 [47.22713366  3.91163927] 45.93751534481224
160 [47.2307617   3.91096933] 45.93750116656743
180 [47.23176204  3.91078461] 45.937500088686626


In [13]:
n = X.shape[1]
alpha = 0.0575
W = np.array([1, 0.5])
W, alpha

(array([1. , 0.5]), 0.0575)

In [14]:
for i in range(200): 
    y_pred = np.dot(W, X)
    err = calc_mse(y, y_pred)    
    for ii in range(W.shape[0]):
        W[ii] -= alpha * (1/n * 2 * np.sum(X[ii] * (y_pred - y)))
    if i % 10 == 0:
        print(i, W, err)

0 [ 7.3365 20.05  ] 3173.15
10 [24.92077178  8.43000744] 354.9322919819926
20 [35.1524411   6.15950171] 134.73705587463752
30 [40.70949545  5.11599852] 71.81301886714115
40 [43.71091805  4.56097305] 53.478178800870616
50 [45.33126     4.26172856] 48.13501639376065
60 [46.20598007  4.10020317] 46.57790366293325
70 [46.67818451  4.01300694] 46.12412743662682
80 [46.93309689  3.96593541] 45.991887259345724
90 [47.07070745  3.94052457] 45.953349620144856
100 [47.14499441  3.92680691] 45.942118921081104
110 [47.1850971   3.91940164] 45.938846053202454
120 [47.20674592  3.91540402] 45.93789226892861
130 [47.21843271  3.91324597] 45.93761431562444
140 [47.22474164  3.91208098] 45.93753331403797
150 [47.22814742  3.91145207] 45.93750970842905
160 [47.22998598  3.91111257] 45.93750282924559
170 [47.2309785   3.91092929] 45.93750082450317
180 [47.2315143   3.91083035] 45.93750024027801
190 [47.23180354  3.91077694] 45.93750007002218


Ответ: скорость обучения - 0.058. Точность до 4 знака после запятой потребует 140 итераций. 


*2. В этом коде мы избавляемся от итераций по весам, но тут есть ошибка, исправьте ее:

In [15]:
n = X.shape[1]
alpha = 0.0575
W = np.array([1, 0.5])
W, alpha

(array([1. , 0.5]), 0.0575)

In [16]:
for i in range(200):
    y_pred = np.dot(W, X)
    err = calc_mse(y, y_pred)
    '''for ii in range(W.shape[0]):
    W[ii] -= alpha * (1/n * 2 * np.sum(X[ii] * (y_pred - y)))'''
    W -= (alpha * (1/n * 2 * np.sum(X * (y_pred - y))))
    if i % 10 == 0:
        print(i, W, err)
        
        
alpha = alpha_new
n_iter = n_iter_new
W = np.array([1, 0.5])
for i in range(n_iter+1):
    y_pred = np.dot(W, X)
    err = calc_mse(y, y_pred)
    '''for ii in range(W.shape[0]):
    W[ii] -= alpha * (1/n * 2 * np.sum(X[ii] * (y_pred - y)))'''
    W -= (alpha * (1/n * 2 * np.dot(X,(y_pred - y))))
    if i % 10 == 0:
        print(i, W, err)

0 [26.8865 26.3865] 3173.15
10 [423.29739408 422.79739408] 1818228.4750305056
20 [11251.04026474 11250.54026474] 1355992545.7139463
30 [307004.80844745 307004.30844745] 1011676755874.8766
40 [8385354.32590033 8385353.82590033] 754790589721282.8
50 [2.29040974e+08 2.29040973e+08] 5.631332650843566e+17
60 [6.25612637e+09 6.25612637e+09] 4.201417963649742e+20
70 [1.70882608e+11 1.70882608e+11] 3.134588915224986e+23
80 [4.66756331e+12 4.66756331e+12] 2.338650368152338e+26
90 [1.27491893e+14 1.27491893e+14] 1.7448174840069977e+29
100 [3.48237008e+15 3.48237008e+15] 1.3017713523812197e+32
110 [9.51189999e+16 9.51189999e+16] 9.712240216602712e+34
120 [2.59812253e+18 2.59812253e+18] 7.246096624606886e+37
130 [7.09662703e+19 7.09662703e+19] 5.406159147853687e+40
140 [1.93840416e+21 1.93840416e+21] 4.0334207844637745e+43
150 [5.29464305e+22 5.29464305e+22] 3.009249779670867e+46
160 [1.44620227e+24 1.44620227e+24] 2.2451374950340191e+49
170 [3.95022099e+25 3.95022099e+25] 1.6750495108980143e+52
1